File: src/controller.py
#!/usr/bin/env python

from __future__ import print_function

import math

import rospy
from geometry_msgs.msg import Twist
from std_msgs.msg import Empty

def stop(pub):
    cmd = Twist()
    pub.publish(cmd)

def forward(x, pub, end_stop=True):
    cmd = Twist()
    cmd.linear.x = 0.2

    start_time = rospy.Time.now()
    # time = start_time
    duration = rospy.Duration.from_sec((x/cmd.linear.x)*1.35) #1.35
    pub.publish(cmd)
    rospy.sleep(duration)
    # while (time - start_time).to_sec() < duration:
    #     pub.publish(cmd)
    #     time = rospy.Time.now()
    #     rate.sleep()

    if end_stop:
        stop(pub)

def rotate(deg, pub, end_stop=True):
    cmd = Twist()
    if deg > 0:
        cmd.angular.z = 0.5
    else:
        cmd.angular.z = -0.5

    theta = abs(deg * math.pi / 180.0)
    start_time = rospy.Time.now()
    # time = start_time
    duration = rospy.Duration.from_sec((theta / 0.5)*1.9) #1.9

    pub.publish(cmd)
    rospy.sleep(duration)

    # while (time - start_time).to_sec() < duration:
    #     pub.publish(cmd)
    #     time = rospy.Time.now()
    #     rate.sleep()

    if end_stop:
        stop(pub)

def curve(pub, end_stop=True):
    cmd = Twist()
    cmd.linear.x = 0.2
    cmd.angular.z = 0.4

    duration = rospy.Duration.from_sec(((math.pi/2.0)/cmd.angular.z)*1.4)
    pub.publish(cmd)
    rospy.sleep(duration)

    if end_stop:
        stop(pub)

def roundBox(pub):
    forward(0.5, pub)
    rospy.sleep(1)
    curve(pub)
    rospy.sleep(1)

    forward(0.5, pub)
    rospy.sleep(1)
    curve(pub)
    rospy.sleep(1)

    forward(0.5, pub)
    rospy.sleep(1)
    curve(pub)
    rospy.sleep(1)

    forward(0.5, pub)
    rospy.sleep(1)
    curve(pub)
    rospy.sleep(1)

def circle(pub):
    cmd = Twist()
    cmd.linear.x = 0.2
    cmd.angular.z = 0.4

    duration = rospy.Duration.from_sec(((math.pi/2.0)/cmd.angular.z)*5)
    pub.publish(cmd)
    rospy.sleep(duration)

    stop(pub)

def boxMovement(pub):
    forward(0.5, pub)
    rospy.sleep(1)
    rotate(90, pub)
    rospy.sleep(1)

    forward(0.5, pub)
    rospy.sleep(1)
    rotate(90, pub)
    rospy.sleep(1)

    forward(0.5, pub)
    rospy.sleep(1)
    rotate(90, pub)
    rospy.sleep(1)

    forward(0.5, pub)
    rospy.sleep(1)
    rotate(90, pub)
    rospy.sleep(1)

def plusMovement(pub):
    forward(0.5, pub)
    rospy.sleep(1)
    rotate(90, pub)
    rospy.sleep(1)
    rotate(90, pub)
    rospy.sleep(1)
    forward(0.5, pub)
    rospy.sleep(1)

    rotate(90, pub)
    rospy.sleep(1)

    forward(0.5, pub)
    rospy.sleep(1)
    rotate(90, pub)
    rospy.sleep(1)
    rotate(90, pub)
    rospy.sleep(1)
    forward(0.5, pub)
    rospy.sleep(1)

    rotate(90, pub)
    rospy.sleep(1)

    forward(0.5, pub)
    rospy.sleep(1)
    rotate(90, pub)
    rospy.sleep(1)
    rotate(90, pub)
    rospy.sleep(1)
    forward(0.5, pub)
    rospy.sleep(1)

    rotate(90, pub)
    rospy.sleep(1)

    forward(0.5, pub)
    rospy.sleep(1)
    rotate(90, pub)
    rospy.sleep(1)
    rotate(90, pub)
    rospy.sleep(1)
    forward(0.5, pub)
    rospy.sleep(1)

    rotate(90, pub)
    rospy.sleep(1)

if __name__ == "__main__":
    rospy.init_node('VO_controller')
    pub = rospy.Publisher('/mobile_base/commands/velocity', Twist, queue_size=1)
    done = rospy.Publisher("/done", Empty, queue_size=1)

    rate = rospy.Rate(100) # 100 Hz

    rospy.sleep(5.)
    # forward(0.5, pub)
    # rotate(-90, pub)
    # curve(pub)
    # boxMovement(pub)
    # plusMovement(pub)
    # roundBox(pub)
    circle(pub)

    done.publish(Empty())

File: src/controller.cpp
#include "ros/ros.h"
#include "std_msgs/Empty.h"
#include "geometry_msgs/Twist.h"

#include <iostream>
#include <math.h>

void stop(ros::Publisher pub);
void forward(ros::Publisher pub, bool end_stop=true);
void rotate(ros::Publisher pub, bool end_stop=true);
void curve(ros::Publisher pub, bool end_stop=true);
void boxMovement(ros::Publisher pub, bool end_stop=true);
void roundBox(ros::Publisher pub, bool end_stop=true);
void circle(ros::Publisher pub, bool end_stop=true);

int main(int argc, char **argv){
  ros::init(argc, argv, "VO_controller");
  ros::NodeHandle n;
  ros::Publisher pub = n.advertise<geometry_msgs::Twist>("/mobile_base/commands/velocity", 1000);
  ros::Publisher done = n.advertise<std_msgs::Empty>("/done", 1000);

  ros::Duration(5.0).sleep();

  rotate(pub);
  rotate(pub);
  //curve(pub);
  //boxMovement(pub);
  //roundBox(pub);
  //circle(pub);

  done.publish(std_msgs::Empty());

  stop(pub);
  ros::Duration(1.0).sleep();

  return 0;
}


void stop(ros::Publisher pub){
  geometry_msgs::Twist cmd;
  pub.publish(cmd);
}

void forward(ros::Publisher pub, bool end_stop){
  ros::WallTime start, end;
  start = ros::WallTime::now();
  geometry_msgs::Twist cmd;
  cmd.linear.x = 0.2;

  ros::Duration duration((0.5/cmd.linear.x)*1.35);
  std::cout << "Forward Duration: " << duration.toSec() << std::endl;
  pub.publish(cmd);
  duration.sleep();

  if(end_stop){
    stop(pub);
    ros::Duration(0.5).sleep();
  }
  end = ros::WallTime::now();
  std::cout << "Forward Execution time: " << (end-start).toSec() << std::endl;
}

void rotate(ros::Publisher pub, bool end_stop){
  ros::WallTime start, end;
  start = ros::WallTime::now();
  geometry_msgs::Twist cmd;
  cmd.angular.z = 0.5;

  ros::Duration duration(((M_PI/2.0)/cmd.angular.z)*2.1);
  std::cout << "Rotate Duration: " << duration.toSec() << std::endl;
  pub.publish(cmd);
  duration.sleep();

  if(end_stop){
    stop(pub);
    ros::Duration(0.5).sleep();
  }
  end = ros::WallTime::now();
  std::cout << "Rotate Execution time: " << (end-start).toSec() << std::endl;
}

void curve(ros::Publisher pub, bool end_stop){
  geometry_msgs::Twist cmd;
  cmd.linear.x = 0.2;
  cmd.angular.z = 0.4;

  ros::Duration duration(((M_PI/2.0)/cmd.angular.z)*1.4);
  pub.publish(cmd);
  duration.sleep();

  if(end_stop){
    stop(pub);
    ros::Duration(1.0).sleep();
  }
}

void boxMovement(ros::Publisher pub, bool end_stop){
  forward(pub, end_stop);
  rotate(pub, end_stop);

  forward(pub, end_stop);
  rotate(pub, end_stop);

  forward(pub, end_stop);
  rotate(pub, end_stop);

  forward(pub, end_stop);
  rotate(pub, true);
}

void roundBox(ros::Publisher pub, bool end_stop){
  ros::Duration wait_time(1.0);

  forward(pub, end_stop);
  wait_time.sleep();
  curve(pub, end_stop);
  wait_time.sleep();

  forward(pub, end_stop);
  wait_time.sleep();
  curve(pub, end_stop);
  wait_time.sleep();

  forward(pub, end_stop);
  wait_time.sleep();
  curve(pub, end_stop);
  wait_time.sleep();

  forward(pub, end_stop);
  wait_time.sleep();
  curve(pub, end_stop);
  wait_time.sleep();
}

void circle(ros::Publisher pub, bool end_stop){
  geometry_msgs::Twist cmd;
  cmd.linear.x = 0.2;
  cmd.angular.z = 0.4;

  ros::Duration duration(((M_PI/2.0)/cmd.angular.z)*5);
  pub.publish(cmd);
  duration.sleep();

  if(end_stop)
    stop(pub);
}
